java-frontend/src/main/java/org/sonar/java/JavaFrontend.java

```diff
@@ -199,20 +202,24 @@ private void scanInBatches(BatchModeContext context, List<InputFile> allInputFil
     }
   }
 
-  private <T extends InputFile> void scanBatch(BatchModeContext context, List<T> allFiles, AnalysisProgress analysisProgress) {
-    analysisProgress.startBatch(allFiles.size());
+  private <T extends InputFile> void scanBatch(BatchModeContext context, List<T> batchFiles, AnalysisProgress analysisProgress) {
+    analysisProgress.startBatch(batchFiles.size());
+    Set<INameEnvironment> environmentsToClean = new HashSet<>();
```

Comment:
This would become:
```java
Set<Runnable> environmentsCleaners = new HashSet<>();
```

I'm however not 100% sure that the `set` would match with method references on same objects. This is to be checked. Meaning, do `c1` and `c2` have the same hashcode, if attached to the same `env` ?
```java
Runnable c1 = env::cleanup;
Runnable c2 = env::cleanup;
```